### ISSUE TITLE
MSSQL engine treats a float literal as a decimal and can truncate its precision

### ISSUE THREAD

**Issue by rebkwok** [opensafely-core/databuilder#1065]
TL;DR: we should add a custom Float type to mssql_dialect, as we do with dates so that float literals are interpreted as floats and not decimal. 
 
Seen in generative tests:
https://github.com/opensafely-core/databuilder/actions/runs/4200368991/jobs/7286312970

As a spec test:
```
table_data = {
        p: """
              | b1 
            --+----
            1 | T    
            """,    
        e: """
              | b1 | f1
            --+----+---
            1 | F  | 0.1 
            """,
    }
    spec_test(
        table_data,
        case(
            when(p.b1).then(  # always False
                case(
                    when(e.b1).then(0.1),
                    default=None
                ).sum_for_patient()            
            ),
            default=0.75
        ),
        {1: 0.75,}
    )
```
This gives a result of 0.8 with the mssql engine, rather than the expected default 0.75.

It seems this is due to literals that we intended to be floats being treated as decimals in mssql, which do not have fixed precision, and so the `sum_for_patient()` in the case statement causes mssql to choose the precision it thinks is most appropriate.  In this case, it's truncating to the same prevision as the `then` clause.
https://learn.microsoft.com/en-us/sql/t-sql/data-types/precision-scale-and-length-transact-sql?redirectedfrom=MSDN&view=sql-server-ver16

This returns the correct result:
```
case(
            when(p.b1).then(  # always False
                case(
                    when(e.b1).then(0.01),  # same precision as default
                    default=None
                ).sum_for_patient()            
            ),
            default=0.75
        ),
```

as does 
```
case(
            when(p.b1).then(  # always False
                case(
                    when(e.b1).then(e.f1),  # non-literal
                    default=None
                ).sum_for_patient()            
            ),
            default=0.75
        ),
```

It's also fine if there is at least one other condition with a non-literal:
```
case(
            when(p.b1).then(  # always False
                case(
                    when(e.b1).then(0.1),
                    when(e.f1 > 0).then(e.f1),
                    default=None
                ).sum_for_patient()            
            ),
            default=0.75
        ),
        {1: 0.75,}
```
#1063 would help this, by simplifying sole constants in case statements.

However, as @evansd [said](https://bennettoxford.slack.com/archives/C03FB777L1M/p1676629911887419?thread_ts=1676608640.640799&cid=C03FB777L1M):
> I think we might need to add a custom Float type to mssql_dialect, as we do with dates. It looks like there's a custom way of formatting literals that ensures they are interpreted as floats rather than decimals:
https://stackoverflow.com/questions/3024306/tsql-make-a-literal-float-value/54974960#54974960

**Comment by evansd** [comment]
On a hunch I tried running this spec test using the pyODBC driver instead of pymssql as we're currently using, and it passed. It was a bit of a faff to install and it wasn't a drop-in replacement (I had to comment out a bunch of the temporary table stuff to get it to work) but I think it demonstrates that the way pymssql handles parameter types is causing us problems here.

I think the same thing was at the root of the date issues we had (#889 and #998) which I worked around here: https://github.com/opensafely-core/databuilder/pull/1008/commits/04989f92954b9dbcd6b63ef0a383cb8b59c0512b

This could be down to differences between the TDS protocol (as used by pymssql) and the ODBC protocol, rather than any deficiency in pymssql.

We initially opted for TDS-based drivers because they seemed to be robust against the problem which caused pyODBC to [die midway](https://github.com/opensafely-core/cohort-extractor/pull/286) through large downloads. And of these, pymssql seemed like the [best option](https://github.com/opensafely-core/cohort-extractor/issues/704) at the time work started on Data Builder.

But in the end, the random disconnects problem didn't entirely go away and so we ended up switching to batched, retry-able downloads. This means we're free to use whichever driver we want, though we stuck with pymssql because it has the nicest installation story.

Perhaps we should reconsider?

However, I think there's an argument that pymsqql is actually working in our favour here. All the bugs we've hit which pyODBC would have handled correctly would have been silently present in the SQL generated by `dump-dataset-sql`. By being "dumb" about types, pymssql forces us to address these issues in a way which fixes the dumped SQL as well. And I think that's a good thing.
